### PR TITLE
Forward compatibility with react/promise 3

### PR DESCRIPTION
**Pull request by WyriHaximus**

### CODE REVIEW

**Comment by clue** [comment]
@WyriHaximus Good initiative and would love to see this getting some traction again! What are your plans for this PR? It looks like all outstanding PRs have been merged in the meantime? (https://github.com/reactphp/dns/pull/153, https://github.com/reactphp/promise-timer/pull/55, https://github.com/reactphp/promise-timer/pull/54, https://github.com/reactphp/promise-timer/pull/37)

**Comment by WyriHaximus** [comment]
@clue https://github.com/reactphp/datagram/actions/runs/3024442084 shows it passing with promise v3 (@dev) working for supported PHP versions.

File: src/Factory.php [hunk 1]
```diff
@@ -140,7 +140,7 @@ function ($_, $reject) use ($promise) {
                 $reject(new \RuntimeException('Cancelled creating socket during DNS lookup'));
 
                 // (try to) cancel pending DNS lookup, otherwise ignoring its results
-                if ($promise instanceof CancellablePromiseInterface) {
+                if (method_exists($promise, 'cancel')) {
```
Comment (clue) [inline src/Factory.php]: Nit: Suggested as per your #28:

```suggestion
                if (\method_exists($promise, 'cancel')) {
```

It also looks like the now unneeded import for `CancellablePromiseInterface` is still in place.

Comment (WyriHaximus) [reply src/Factory.php]: My bad, thought I removed that :+1:

**Review by clue — APPROVED** [review]
@WyriHaximus Thanks for the update, changes LGTM! :shipit:

**Pull request merged**